block_windows.go

```diff
@@ -0,0 +1,233 @@
+
+type win32LogicalDisk struct {
+	Caption           string
+	CreationClassName string
+	Description       string
+	DeviceID          string
+	FileSystem        string
+	FreeSpace         uint64
+	Name              string
+	Size              uint64
+	SystemName        string
+}
+
+func (ctx *context) blockFillInfo(info *BlockInfo) error {
+	win32DiskDriveDescriptions, err := getDiskDrives()
+	if err != nil {
+		return err
+	}
+
+	win32DiskPartitionDescriptions, err := getDiskPartitions()
+	if err != nil {
+		return err
+	}
+
+	win32LogicalDiskToPartitionDescriptions, err := getLogicalDisksToPartitions()
+	if err != nil {
+		return err
+	}
+
+	win32LogicalDiskDescriptions, err := getLogicalDisks()
+	if err != nil {
+		return err
+	}
+
+	// Converting into standard structures
+	disks := make([]*Disk, 0)
+	for _, diskdrive := range win32DiskDriveDescriptions {
+		disk := &Disk{
+			Name:                   diskdrive.Name,
+			SizeBytes:              diskdrive.Size,
+			PhysicalBlockSizeBytes: 0,
+			DriveType:              toDriveType(diskdrive.MediaType),
+			StorageController:      toStorageController(diskdrive.InterfaceType),
+			BusType:                toBusType(diskdrive.InterfaceType),
+			BusPath:                UNKNOWN, // TODO: add information
+			Vendor:                 UNKNOWN, // TODO: add information
+			Model:                  diskdrive.Caption,
+			SerialNumber:           diskdrive.SerialNumber,
+			WWN:                    UNKNOWN, // TODO: add information
+			Partitions:             make([]*Partition, 0),
+		}
+		for _, diskpartition := range win32DiskPartitionDescriptions {
+			// Finding disk partition linked to current disk drive
+			if diskdrive.Index == diskpartition.DiskIndex {
+				disk.PhysicalBlockSizeBytes = diskpartition.BlockSize
+				// Finding logical partition linked to current disk partition
+				for _, logicaldisk := range win32LogicalDiskDescriptions {
+					for _, logicaldisktodiskpartition := range win32LogicalDiskToPartitionDescriptions {
+						var desiredAntecedent = "\\\\" + diskpartition.SystemName + "\\root\\cimv2:" + diskpartition.CreationClassName + ".DeviceID=\"" + diskpartition.DeviceID + "\""
+						var desiredDependent = "\\\\" + logicaldisk.SystemName + "\\root\\cimv2:" + logicaldisk.CreationClassName + ".DeviceID=\"" + logicaldisk.DeviceID + "\""
+						if logicaldisktodiskpartition.Antecedent == desiredAntecedent && logicaldisktodiskpartition.Dependent == desiredDependent {
+							// Appending Partition
+							p := &Partition{
+								Name:       logicaldisk.Caption,
+								Label:      logicaldisk.Caption,
+								SizeBytes:  logicaldisk.Size,
+								MountPoint: logicaldisk.DeviceID,
+								Type:       diskpartition.Type,
+								IsReadOnly: toReadOnly(diskpartition.Access), // TODO: add information
+							}
+							disk.Partitions = append(disk.Partitions, p)
+							break
+						}
+					}
+				}
+			}
+		}
```

Comment:
I have an idea on how we can remove the above Go code by just using a JOIN in the WQL statement... but we can work on that together in a followup patch :)